langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Reverse.java

```diff
@@ -18,23 +18,82 @@
 
 package org.ballerinalang.langlib.array;
 
+import io.ballerina.runtime.api.TypeTags;
+import io.ballerina.runtime.api.creators.TypeCreator;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.types.ArrayType;
+import io.ballerina.runtime.api.types.TupleType;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.types.UnionType;
 import io.ballerina.runtime.api.values.BArray;
 
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import static org.ballerinalang.langlib.array.utils.ArrayUtils.createOpNotSupportedError;
+
 /**
  * Native implementation of lang.array:reverse((any|error)[]).
  *
  * @since 1.0
  */
-//@BallerinaFunction(
-//        orgName = "ballerina", packageName = "lang.array", functionName = "reverse",
-//        args = {@Argument(name = "arr", type = TypeKind.ARRAY)},
-//        returnType = {@ReturnType(type = TypeKind.ARRAY)},
-//        isPublic = true
-//)
 public class Reverse {
 
     public static BArray reverse(BArray arr) {
-        arr.reverse();
-        return arr;
+        Type arrType = arr.getType();
+        BArray revArr;
+        switch (arrType.getTag()) {
+            case TypeTags.ARRAY_TAG:
+                revArr = (BArray) arr.copy(new HashMap<>());
+                revArr.reverse();
+                return revArr;
+            case TypeTags.TUPLE_TAG:
+                revArr = createArrayFromTuple(arr);
+                revArr.reverse();
+                return revArr;
+            default:
+                throw createOpNotSupportedError(arrType, "reverse()");
+        }
+    }
+
+    private static BArray createArrayFromTuple(BArray arr) {
+        Type arrType = arr.getType();
+        TupleType tupleType = (TupleType) arrType;
+        List<Type> memTypes = new ArrayList<>();
+        List<Type> tupleTypes = tupleType.getTupleTypes();
+        boolean isSameType = true;
+        Type sameType = null;
+        if (!tupleTypes.isEmpty()) {
+            sameType = tupleTypes.get(0);
+            memTypes.add(sameType);
+        }
+        for (int i = 1; i < tupleTypes.size(); i++) {
+            isSameType &= sameType == tupleTypes.get(i);
+            memTypes.add(tupleTypes.get(i));
+        }
+        Type restType = tupleType.getRestType();
+        if (restType != null) {
+            isSameType = false;
+            memTypes.add(restType);
```

Comment:
create empty array from tuple has been duplicated. Can we have a common method. We can put into ArrayUtils class